crates/fuel-core/src/graphql_api/required_fuel_block_height_extension.rs

```diff
@@ -0,0 +1,149 @@
+pub(crate) struct RequiredFuelBlockHeightExtension;
+
+impl RequiredFuelBlockHeightExtension {
+    pub fn new() -> Self {
+        Self
+    }
+}
+
+impl ExtensionFactory for RequiredFuelBlockHeightExtension {
+    fn create(&self) -> Arc<dyn Extension> {
+        Arc::new(RequiredFuelBlockHeightExtension::new())
+    }
+}
+
+pub(crate) struct RequiredFuelBlockHeightTooFarInTheFuture;
+
+#[async_trait::async_trait]
+impl Extension for RequiredFuelBlockHeightExtension {
+    async fn prepare_request(
+        &self,
+        ctx: &ExtensionContext<'_>,
+        request: Request,
+        next: NextPrepareRequest<'_>,
+    ) -> ServerResult<Request> {
+        let database: &ReadDatabase = ctx.data_unchecked();
+        let required_fuel_block_height_header_value = request
+            .extensions
+            .get(REQUIRED_FUEL_BLOCK_HEIGHT_HEADER)
+            .map(|value| match value {
+                async_graphql::Value::Number(number) => {
+                    // Safety: The value was constructed
+                    BlockHeight::new(
+                        number
+                            .as_u64()
+                            .and_then(|n| n.try_into().ok())
+                            .expect("The REQUIRED_FUEL_BLOCK_HEIGHT_HEADER value has been constructed from a u64 value"),
+                    )
+                }
+                _ => panic!("The REQUIRED_FUEL_BLOCK_HEIGHT_HEADER value has been constructed from a u64 value"),
+            });
+
+        let latest_known_block_height = database
+            .view()
+            .and_then(|view| view.latest_block_height())
+            .map_err(|e| {
+                let (line, column) = (line!(), column!());
+                ServerError::new(
+                    e.to_string(),
+                    Some(Pos {
+                        line: line as usize,
+                        column: column as usize,
+                    }),
+                )
+            })?;
+
+        {
+            // At this point, the query_data in the ExtensionContext is empty.
+            // See https://github.com/async-graphql/async-graphql/blob/7f1791488463d4e9c5adcd543962173e2f6cbd34/src/schema.rs#L521
+            // We need to fetch the mutable location to store the current fuel block height
+            // directly from request.data
+            let mut current_block_height = request
+            .data
+            .get(&TypeId::of::<Arc<Mutex<Option<BlockHeight>>>>())
+            .and_then(|data| { println!("Request was here"); data.downcast_ref::<Arc<Mutex<Option<BlockHeight>>>>()})
+            .expect("Data to store current fuel block height was set when forming the request")
+            .lock()
+            .await;
+
+            *current_block_height = Some(latest_known_block_height);
+        }
+
+        if let Some(required_fuel_block_height) = required_fuel_block_height_header_value
+        {
+            if required_fuel_block_height > latest_known_block_height {
```

Comment:
todo: implement From<RequiredFuelBlockHeightTooFarInTheFuture> for ServerError

Reply:
edit: I decided not to, as the final code seemed to be less readable than the current one